### ISSUE TITLE
Incorrect calculation of PPO critic loss

### ISSUE THREAD

**Issue by alexaoliveira2000** [nikhilbarhate99/PPO-PyTorch#72]
### **Problem**
According to equation 9 in the original PPO paper ([link](https://arxiv.org/abs/1707.06347)), the objective is to maximize:
$L(\theta) = \mathbb{E} \left[ L^{\text{clip}}(\theta) - c_1  L^{\text{VF}}(\theta) + c_2 S\[\pi_\theta\](s_t) \right]$,
where the critic loss is given by a squared error (not mean-squared error!). That is, $L^{\text{VF}}(\theta) = (V_\phi(s_t) - R_t)^2$. The **mean** part naturally appears in the equation because of the expected value.

In your code, however, you're considering $L^{\text{VF}}(\theta) = \frac{1}{|B|} \sum_{s_t \sim B} (V_\phi(s_t) - R_t)^2$:
`loss = -torch.min(surr1, surr2) + 0.5 * self.MseLoss(state_values, rewards) - 0.01 * dist_entropy`
to which you are then applying the mean over samples (before backprop): `loss.mean().backward()`,

**Is this really a problem in optimization?**
If you consider $L^{\text{VF}}$ as an MSE, then the gradient is effectively the same (it is learning in the direction it was intended) but gradients are scaled down (learning is slower). To scale up again, you could multiply $c_1$ by the number of samples (n), that is, you could do something like:
`loss = -torch.min(surr1, surr2) + n * 0.5 * self.MseLoss(state_values, rewards) - 0.01 * dist_entropy`
but this seems ugly.

### **Solution**
In the PPO loss line, just change the critic loss to not apply mean reduction, so it represents the squared error (per sample error), without the mean:
`loss = -torch.min(surr1, surr2) + 0.5 * self.MseLoss(state_values, rewards, reduction='none') - 0.01 * dist_entropy`

This way, each of the components (actor loss, critic loss and entropy bonus) represents a vector where the length of each is the number of samples.
The expected value part is already implemented by doing the `loss.mean()`.

### **Closing Remarks**
Btw, I really like the implementation and have been using it for months!
